src/mds/Server.cc

```diff
@@ -2334,18 +2334,20 @@ void Server::set_trace_dist(const ref_t<MClientReply> &reply,
     encode(dn->get_name(), bl);
 
     int lease_mask = 0;
-    CDentry::linkage_t *dnl = dn->get_linkage(mdr->get_client(), mdr);
-    if (dnl->is_primary()) {
-      ceph_assert(dnl->get_inode() == in);
-      lease_mask = CEPH_LEASE_PRIMARY_LINK;
-    } else {
-      if (dnl->is_remote())
-	ceph_assert(dnl->get_remote_ino() == in->ino());
-      else
-	ceph_assert(!in);
+    if (!dn->is_projected() || dn->use_projected(mdr->get_client(), mdr)) {
```

Comment:
@vshankar Is this check allowing if it's not projected? I don't understand it :)

Reply:
MDCache::path_traverse() fills in mds->dn[0] based on a dentry's projected linkage. When responding to client, checks in Server::set_trace_dist() might not use projected linkages depending on the client (dn->get_linkage(mds->get_client(), mdr)).